File: Tree/144.Binary-Tree-Preorder-Traversal.py
# https://leetcode.com/problems/binary-tree-preorder-traversal/

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    
    def preorderTraversal(self, root: TreeNode) -> List[int]:
        result = []
        node: TreeNode = root
        
        while node is not None:
            if node.left is None:
                result.append(node.val)
                node = node.right
            else:
                predecessor = node.left # node cuoi cung bi duyet trong left subtree of node        
                
                while predecessor.right is not None and predecessor.right is not node:
                    predecessor = predecessor.right
                
                if predecessor.right is None: # day la lan dau duyet qua node 
                    result.append(node.val)
                    predecessor.right = node
                    node = node.left
                else:
                    predecessor.right = None
                    node = node.right
                    
        
        return result

File: Tree/144.Binary-Tree-Preorder-Traversal.java
/*
https://leetcode.com/problems/binary-tree-preorder-traversal/
*/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    // root left right
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        TreeNode curr = root;
        if(root == null) {
            return result;
        }
        Stack<TreeNode> stack = new Stack<>();
        stack.push(root);
        while(!stack.isEmpty()) {
            TreeNode currTop = stack.pop();
            result.add(currTop.val);
            
            if(currTop.right != null) {
                stack.push(currTop.right);
            }
            
            if(currTop.left != null) {
                stack.push(currTop.left);
            }
            
        }
        
        return result;
    }
}
